### PR TITLE
feat: add wds modal component

### LINKED ISSUE #28463: Create WDS modal component
We only make a paper model with sizes, adding a header, footer and content area is Anvil team responsibility.

Example of reusable modal component — https://floating-ui.com/docs/dialog 

Modal should have 3 semantic options for size:
- info
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBLzEvQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--e4279068350bda3e6a87e15fbac3c2d3373e1391/CleanShot%202023-10-25%20at%2018.30.08.png)

- contentFull
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBLzUvQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--57683ef33b70bc273e66f544162d66350a16a83a/CleanShot%202023-10-25%20at%2018.31.16.png) 

- fullScreen 
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBd1NBQVE9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--1e564323468d1de0ff057a6f16b25a63744915da/CleanShot%202023-10-27%20at%2012.38.23.png)

### PR DESCRIPTION
**Pull request by KelvinOm**
## Description
Add WDS Modal component

#### PR fixes following issue(s)
Fixes #28463 

#### Media
https://github.com/appsmithorg/appsmith/assets/11555074/38871bdb-32ec-48c6-878d-75c831057b69

#### Type of change
- New feature (non-breaking change which adds functionality)

- [x] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### CODE REVIEW

File: app/client/packages/design-system/headless/src/components/Button/src/Button.tsx [hunk 1]
```diff
@@ -26,16 +26,21 @@ const _Button = (props: ButtonProps, ref: ButtonRef) => {
     className,
     draggable = false,
     isDisabled = false,
+    ...rest
   } = props;
   const { hoverProps, isHovered } = useHover({ isDisabled });
   const { focusProps, isFocusVisible } = useFocusRing({ autoFocus });
-  const { buttonProps, isPressed } = useButton(props, ref as ButtonRefObject);
+  const { buttonProps, isPressed } = useButton(
```
Comment (KelvinOm) [inline app/client/packages/design-system/headless/src/components/Button/src/Button.tsx:33]: fix button autofocus

File: app/client/packages/design-system/theming/src/theme/stories/Spacing.stories.mdx [hunk 2]
```diff
@@ -16,26 +16,15 @@ spacing system can add clarity and consistency to a design.
...
-### Outer Spacing
-
-Outer spacing, also known as **margin** and **gap**, is the space
-between elements. It creates a visual separation between elements and
-contributes to the overall balance of a design.
-
-## Tokens Table
-
-export const Spacing = () => {
+export const Spacing = ({ type }) => {
```
Comment (KelvinOm) [inline app/client/packages/design-system/theming/src/theme/stories/Spacing.stories.mdx:19]: fix tokens table in spacing story

File: app/client/packages/storybook/.storybook/styles.css [hunk 3]
```diff
@@ -27,3 +27,7 @@ body,
 .is-chromatic * {
   animation: none !important;
 }
+
+*:focus {
```
Comment (KelvinOm) [inline app/client/packages/storybook/.storybook/styles.css:31]: fix for focus from blueprint styles
Снимок экрана 2023-11-08 в 16 36 05

**Review by ichik — APPROVED** [review]

File: app/client/packages/design-system/widgets/src/components/Modal/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  /** Close button text
+   * @default Close
+   */
+  closeText?: string;
+  /** The event that is triggered when the submit button is clicked. */
+  onSubmit?: () => void;
+}
+
+export interface ModalBodyProps {
+  children: ReactNode;
```
Comment (marks0351) [inline app/client/packages/design-system/widgets/src/components/Modal/src/types.ts:31]: @KelvinOm to open and close the widget should i just create and destroy the component? should we add a prop to toggle this state? 

also would need a callback when the widget closes which i think we should be able to handle on the widget side also.

also how would it absorb a custom background color?

**Review by dhruvikn — COMMENTED** [review]
Apart from @marks0351's comment, everything looks fine to me. 
Moreover, thank you for adding more test cases. 💯

File: app/client/packages/design-system/widgets/src/components/Modal/src/types.ts (hunk 4 above)
Comment (KelvinOm) [reply app/client/packages/design-system/widgets/src/components/Modal/src/types.ts:37]: @marks0351 Thanks for the review

>to open and close the widget should i just create and destroy the component? should we add a prop to toggle this state?

Added props to manage the opening. Also note that the modal can work in two modes (controlled and uncontrolled)

**Controlled**
```
export const ControlledModal = () => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <Modal isOpen={isOpen} setOpen={setIsOpen}>
      <Button>My modal trigger</Button>
      <ModalBody>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Doloribus,
        vero!
      </ModalBody>
    </Modal>
  );
};
```

**Uncontrolled**
```
    <Modal>
      <Button>My modal trigger</Button>
      <div>My modal content</div>
    </Modal>
```

>also would need a callback when the widget closes which i think we should be able to handle on the widget side also.
Added `onClose` callback.

> also how would it absorb a custom background color?
There should be no custom styles, all styling should be inherited from the theme. But in anyway, I added the ability to pass classes for content and overlay.

All the changes described above are reflected in the storybook documentation.

**Review by ichik — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Pull request merged**